piptools/scripts/compile.py

```diff
@@ -86,6 +92,9 @@ def cli(verbose, dry_run, pre, rebuild, find_links, index_url, extra_index_url,
         if not output_file:
             raise click.BadParameter('--output-file is required if input is from stdin')
 
+    if any(map(os.path.isdir, src_files)) and not output_file:
+        raise click.BadParameter('--output-file is required if an input directory is given.')
```

Comment:
I was hoping for a different behaviour: for each `.in` file found in the directory (or directories), use `{basename}.txt` as output file.  The idea is to have different files for runtime requirements (direct dependencies), test deps (additional things for CI, not installed during deployment), extra deployment deps (not installed locally or duing CI).  At the moment I have a wrapper script that iterates over `in` files to create separate `txt` files.

Reply:
@merwok If I'm following your use correctly, you would have something like:
requirements.in:
```
myapp
requests[security]
```
and then test-requirements.in:
```
pytest
flake8
```
and you would want separate requirements.txt and test-requirements.txt, which are the `pip-compile` output of each?
If so, then you might have duplicate dependencies in each, which would force you to do `pip install` of each separately, but that breaks the guarantee that the dependencies of each will be met (your test-requirements.txt and requirements.txt `requests` requirement may differ, as an example).

Reply:
This can be solved, if you define which of the files is the "master".

I have solved this in Prequ by making the main `requirements.txt` as the master which is always compiled first. When compiling the other requirements files, Prequ will add
```
-c requirements.txt
```
line to the input lines so that the compiled `requirements-*.txt` won't have any conflicting dependencies. Here's the relevant code: https://github.com/suutari/prequ/blob/v1.3.0/prequ/configuration.py#L182

Though, there might still be conflicting dependencies in requirements-a.txt and requirements-b.txt, if those are not in requirements.txt, but that's not very common case (and it's not yet handled by Prequ).

Reply:
@suutari-ai Thank you for the input. I like where you're going with this. I will have to revisit this idea after my coffee, but it seems to make sens to me right now.

@simlun Did you have a specific use case for the way you implemented this, or where you simply (and kindly!) taking care of an existing open issue?

Let's make sure we're all on the same page 😄

Reply:
@vphilippon Yes that’s right.  I don’t have issues with duplicate dependencies; typically I have Pyramid or Django and various libs in the runtime requirements, test deps that work with Pyramid or Django but don’t depend on them (see https://github.com/nedbat/django_coverage_plugin/issues/44 for example), CI dependencies which are an other disjoint set (tox, coverage and pip-tools).  To run tests I call `pip-sync runtime.txt tests.txt` (see https://github.com/tox-dev/tox/issues/149#issuecomment-290263985 ), and the only problem I’ve had was with six, which I solved by re-compiling the files so that the latest version gets picked up in both files.  So far I haven’t used test deps that need requests, which could easily cause a version conflict since third-party API clients will probably depend on a specific version range of requests.

This has been working very well on many projects in the last year!  Depending on the task, different sets are installed: ci.txt to bootstrap, runtime.txt + tests.txt inside tox testenvs, runtime.txt + deploy.txt for deployment.  I was looking forward to this ticket to be able to do `pip-compile deps/` and remove my wrapper script.

If it is agreed that this is about taking many sources and creating a unique frozen requirements file, then I’ll probably open another ticket to ask for an option to make pip-compile accept multiple source files (or a directory) and compile each one to its matching frozen version.  Then the two use cases would be satisfied :)

Reply:
@vphilippon I just love this project and was simply supplying a PR for an existing open issue. If I misunderstood, or someone wants to take it in another direction - feel free! :)